### PR TITLE
Fix OkHttpURLConnection to not swallow unexpected exceptions.

### PR DESCRIPTION
**Pull request by swankjesse**
Previously we would sometimes delegate to the asynchronous implementation,
and that would prevent exceptions from reaching their destination.

### CODE REVIEW

**Comment by swankjesse** [comment]
. . . the underlying problem is that `Callback.onFailure()` expects an IOException. If any other kind of exception prevents a response from being delivered, `Callback.onFailure()` is never called. This is generally legit – the program has crashed! But in this case it’s surprising that a synchronous call never receives a response.

File: okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java [hunk 1]
```diff
@@ -587,6 +588,32 @@ private void setProtocols(String protocolsString, boolean append) {
...
   }
 
+  static final class UnexpectedException extends IOException {
+    static final Interceptor INTERCEPTOR = new Interceptor() {
+      @Override public Response intercept(Chain chain) throws IOException {
+        try {
+          return chain.proceed(chain.request());
+        } catch (IOException e) {
+          throw e;
+        } catch (Error | RuntimeException e) {
```
Comment (dave-r12) [inline okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java]: Good catch, literally! I can't say I've caught an Error before, I assume that's ok.

File: okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java [hunk 2]
```diff
@@ -587,6 +588,32 @@ private void setProtocols(String protocolsString, boolean append) {
...
+      }
+    };
+
+    public UnexpectedException(Throwable cause) {
+      super(cause);
+    }
+  }
+
+  /** Throws {@code throwable} as either an IOException, RuntimeException, or Error. */
+  private static IOException propagate(Throwable throwable) throws IOException {
```
Comment (dave-r12) [inline okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java]: Should this to be a void return type?

File: okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java (hunk 1 above)
Comment (swankjesse) [inline okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java]: Yeah, it’s most useful to catch an AssertionError . . . it’ll get rethrown shortly and hopefully not caught!

File: okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java (hunk 2 above)
Comment (swankjesse) [inline okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java]: The unused return type exists to make the callsites happy . . . 
https://publicobject.com/2008/05/calling-method-that-always-throws.html

Comment (dave-r12) [inline okhttp-urlconnection/src/main/java/okhttp3/internal/huc/OkHttpURLConnection.java]: Oh, good read, thanks.

**Comment by dave-r12** [comment]
Nice, lgtm.

**Pull request merged**